### PR TITLE
Refactor Recent Events & Add to Org Box- SEAB-5267

### PR DESCRIPTION
**Pull request by KimberleyChong**
**Description**
Adds recent events to the `organization-box` (SEAB-5267), refactors the `recent-events.component` so that it can be  used to display recent events for the `organization-box.component`, `starred-box.component` and `user-page.component`, and adds the `recent-events.pipe` to assist in cleaning up the `recent-events` template (SEAB-4056 SEAB-4715 SEAB-4714).

**Org-box & Starred-box**
image
**User Pages**
image

Note that the [search bar](https://app.zeplin.io/project/5fc690a2717ee9144814f190/screen/6022ab2d42df9f2c57c20982) in the organization box was removed as it doesn't make much sense to implement (at least right now), the mocks don't show how search is handled and not sure how much use users would get out of it given that not many belong to an organization. The "All Organizations" link at the bottom of the card is also missing as the `my-organizations` page does not exist.

This pr also cleans up two console errors that would occur due to the use of `ng-plural` in the `org-box`.

### recent-events.component
Now takes an `eventType` parameter to display organization events for the `org-box` or starred events for the `starred-box`. If no input is provided and it identifies a username in the url (i.e. on user page) then it will retrieve the user's events.
https://github.com/dockstore/dockstore-ui2/blob/44df1f0e8432d2d7320dfe6abb432c62d26bbed2/src/app/home-page/recent-events/recent-events.component.ts#L28

The `recent-events.component` now covers `organization events`, `starred events`, and `entry events` covering these event types:

| CREATEORG | MODIFYORG | APPROVEORG | REJECTORG |
| :---| :---| :---| :---| 
| ADDUSERTOORG | APPROVEORGINVITE | CREATECOLLECTION | MODIFYCOLLECTION |
| DELETECOLLECTION | REMOVEFROMCOLLECTION | ADDTOCOLLECTION | ADDVERSIONTOENTRY | 
| PUBLISHENTRY | UNPUBLISHENTRY |

Notable changes to UI:
- The username and user avatar now link to that user's user page, this wasn't a requirement, but after interacting with the component I thought it made sense to have that functionality
- Events that display an org image where that organization does not have an org logo will display the org icon
image
- Entry events (`ADDVERSIONTOENTRY`, `PUBLISHENTRY`, `UNPUBLISHENTRY`) will display the associated entry type icon
- Org icons and entry icons will link to the associated organization or entry unless the event was `UNPUBLISHENTRY` in which the entry would no longer exist
- Likewise, when the event is an `UNPUBLISHENTRY` or `DELETECOLLECTION` the entry/collection name will not have a link and it will appear bolded instead, these checks help prevent some 404 issues
image

### recent-events.pipe
Cleans up these conditionals that would occur in the `recent-events.component` and `starred-box.component`:

Getting an entry link:
```
[routerLink]="
                event?.apptool
                  ? '/tools/' + event?.apptool?.full_workflow_path
                  : event?.tool
                  ? '/tools/' + event?.tool?.tool_path
                  : '/workflows/' + event?.workflow?.full_workflow_path
```
Getting an entry type:
```
{{ event?.tool || event?.apptool ? 'tool' : 'workflow' }}
```
Getting an entry name:
```
{{ event?.apptool ? event?.apptool.workflowName : event?.tool ? event?.tool.name : event?.workflow.workflowName }}
```
Getting an entry display name:
```
{{ (event?.tool | entryToDisplayName) || (event?.workflow | entryToDisplayName) || (event?.apptool | entryToDisplayName) }}
```
The pipe takes in an `Event` object and extracts details about the event given the desired `type` param.
https://github.com/dockstore/dockstore-ui2/blob/44df1f0e8432d2d7320dfe6abb432c62d26bbed2/src/app/shared/entry/recent-events.pipe.ts#L13-L20

**Side note:**
The `starred-box` seems to flash the empty state "You do not have any starred entries" on page load, there was previously an `isLoading` check that would help prevent this issue, but events are now loaded in the `recent-events.component` so this check was removed.  If anyone has any ideas on how to prevent the flash on page load please let me know :)
https://github.com/dockstore/dockstore-ui2/blob/616bf4c69f58c81598232d97d17cf82cb8a6ad48/src/app/home-page/widget/starred-box/starred-box.component.html#L37

**Lastly:**
Here's a little overview of some the new navigation to user pages, icons and org links:

https://user-images.githubusercontent.com/97123241/226028736-e6b26ed3-adc1-4929-a1a8-93a732fa78c2.mp4

**Review Instructions**

**Issue**
SEAB-5267 SEAB-4056 SEAB-4715 SEAB-4714

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that your code compiles by running `npm run build`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] If this is the first time you're submitting a PR or even if you just need a refresher, consider reviewing our [style guide](https://github.com/dockstore/dockstore/wiki/Dockstore-Frontend-Opinionated-Style-Guide#pr-checklist)
- [x] Do not bypass Angular sanitization (bypassSecurityTrustHtml, etc.), or justify why you need to do so
- [x] If displaying markdown, use the `markdown-wrapper` component, which does extra sanitization
- [x] Do not use cookies, although this may change in the future
- [x] Run `npm audit` and ensure you are not introducing new vulnerabilities
- [x] Do due diligence on new 3rd party libraries, checking for CVEs
- [x] Don't allow user-uploaded images to be served from the Dockstore domain
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.
- [x] Check whether this PR disables tests. If it legitimately needs to disable a test, create a new ticket to re-enable it in a specific milestone.

### CODE REVIEW

**Comment by KimberleyChong** [comment]
> Kudos, SonarCloud Quality Gate passed!    [Quality Gate passed](https://sonarcloud.io/dashboard?id=dockstore_dockstore-ui2&pullRequest=1735)
> 
> [Bug](https://sonarcloud.io/project/issues?id=dockstore_dockstore-ui2&pullRequest=1735&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=dockstore_dockstore-ui2&pullRequest=1735&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=dockstore_dockstore-ui2&pullRequest=1735&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=dockstore_dockstore-ui2&pullRequest=1735&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=dockstore_dockstore-ui2&pullRequest=1735&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=dockstore_dockstore-ui2&pullRequest=1735&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=dockstore_dockstore-ui2&pullRequest=1735&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=dockstore_dockstore-ui2&pullRequest=1735&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=dockstore_dockstore-ui2&pullRequest=1735&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=dockstore_dockstore-ui2&pullRequest=1735&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=dockstore_dockstore-ui2&pullRequest=1735&resolved=false&types=CODE_SMELL) [3 Code Smells](https://sonarcloud.io/project/issues?id=dockstore_dockstore-ui2&pullRequest=1735&resolved=false&types=CODE_SMELL)
> 
> [No Coverage information](https://sonarcloud.io/component_measures?id=dockstore_dockstore-ui2&pullRequest=1735) No Coverage information [](https://sonarcloud.io/component_measures?id=dockstore_dockstore-ui2&pullRequest=1735&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=dockstore_dockstore-ui2&pullRequest=1735&metric=new_duplicated_lines_density&view=list)

Will fix with pr feedback

**Comment by kathy-t** [comment]
> Note that the [search bar](https://app.zeplin.io/project/5fc690a2717ee9144814f190/screen/6022ab2d42df9f2c57c20982) in the organization box was removed as it doesn't make much sense to implement (at least right now), the mocks don't show how search is handled and not sure how much use users would get out of it given that not many belong to an organization. The "All Organizations" link at the bottom of the card is also missing as the my-organizations page does not exist.

From my understanding, this PR makes the organization box only have Recent Activity and it doesn't list the organizations that the user belongs to. My concern is that since we're switching to the new dashboard by default, the user won't have a way to view all of their organizations without the My Organizations page, which currently has a 1.15 milestone in its [epic](https://ucsc-cgl.atlassian.net/browse/SEAB-5041). With this PR, is there a way for the user to view their organizations? If not, I would consider at least keeping the My Organizations part of the box.

**Comment by KimberleyChong** [comment]
> With this PR, is there a way for the user to view their organizations? If not, I would consider at least keeping the My Organizations part of the box.

Yeah this is  true, there is currently no way to view a list of your organizations... I had an idea to maybe add a dropdown to link a user's orgs (roughly like this):
image

But there was no strong desire to keep it @denis-yuen

File: src/app/home-page/recent-events/recent-events.component.html
Comment (svonworl) [inline src/app/home-page/recent-events/recent-events.component.html:28]: This is better than before, but I wonder if we could further simplify by either:
1. creating templates for the various "categories" of events, and for each event type, pass the appropriate user/verb/phrase/etc to them.
2. creating some pipes to construct linked text to collections/orgs/events, eliminating much of the duplicate code.

File: src/app/home-page/recent-events/recent-events.component.html
Comment (svonworl) [inline src/app/home-page/recent-events/recent-events.component.html:171]: One failure mode is when the organization specifies an `avatarUrl`, but the image pointed at by the URL doesn't exist or is not valid/loadable.  In that case, to prevent a "broken" icon from being displayed, we implemented some fallback code to load a default image.  We should probably include it here.  See the `appFallback` directive at this line: https://github.com/dockstore/dockstore-ui2/blob/d4f5ccb4c3c1c9ac4be2f4cf72e40f7a33277948/src/app/organizations/organizations/organizations.component.html#L69

File: src/app/shared/entry/recent-events.pipe.ts
Comment (coverbeck) [inline src/app/shared/entry/recent-events.pipe.ts:9]: Typo in accommodate :)

File: src/app/shared/entry/recent-events.pipe.ts
Comment (KimberleyChong) [reply src/app/shared/entry/recent-events.pipe.ts:21]: Mm got it, thanks!

File: src/app/home-page/recent-events/recent-events.component.html
Comment (KimberleyChong) [reply src/app/home-page/recent-events/recent-events.component.html:171]: Works great, thanks Steve :)

**Comment by KimberleyChong** [comment]
> So I think both from simplicity and that, I think leaving "my organizations" in the lower part of the widget works.

Don't know how I feel about it... I think it looks a little confused.  This example shows two orgs, but we'd definetly need to impose a display limit so the card doesn't expand too large (2 is probably the max). Feel like it doesn't provide the best functionality like this

image

I think a drop-down is the least visually _invasive_ solution, but open to ideas @kathy-t @coverbeck @svonworl @denis-yuen

**Comment by kathy-t** [comment]
> I think a drop-down is the least visually _invasive_ solution, but open to ideas @kathy-t @coverbeck @svonworl @denis-yuen

Brainstorming, how about a button somewhere in the box that says "View my organizations" and clicking it opens a dialog box that lists the organizations? It gives you a bit more space to work with compared to the drop-down

File: src/app/home-page/recent-events/recent-events.component.html
Comment (KimberleyChong) [reply src/app/home-page/recent-events/recent-events.component.html:28]: Was thinking about this, I think option 1 would be the most viable, would probably require creating a component for entry events, org events and collection events, and we would need to pass the `event` to the template as well to generate the links.  Some cases have additional information that we'd need to also have `*ngIf`s for.

Going to hold off on it because it would be a large refactor I'd prefer to do after this is merged, but will make a ticket for it.

**Comment by KimberleyChong** [comment]
**Updates:**
- Added the `orgLogoService` image fallback to `recent-events.component` for broken org logo images
- Updated `RecentEventsPipe` to use `EntryToDisplayNamePipe` through injection
- Added `MyOrganizationsDialogComponent` used by `OrganizationBoxComonent` to display a list of the user's organizations:


image

https://user-images.githubusercontent.com/97123241/227323411-d9fbf068-a13e-444c-8a92-d822e0bc136c.mp4

Ran into an interesting behaviour where you wouldn't start at the top of the page after navigating from a dialog, see [here](https://github.com/dockstore/dockstore/issues/5438) for details. I resolved the issue by passing the `orgName` back to the parent on dialog close and have the parent navigate to the link instead.

https://github.com/dockstore/dockstore-ui2/blob/5e219764b25e39a11994adc265387484a263a22e/src/app/home-page/widget/organization-box/my-organizations-dialog.component/my-organizations-dialog.component.ts#L15-L18

Left a note in the `OrganizationBoxComponent` to remove the dialog once the `my-organizations` page is implemented.

File: src/app/home-page/recent-events/recent-events.component.ts
Comment (coverbeck) [inline src/app/home-page/recent-events/recent-events.component.ts:119]: This is querying the Akita store, not making an http(s) request. It looks like we generally don't need to handle the error condition for querying the store, since it's just fetching from an in memory object.

Also, you're supposed to unsubscribe. The preferred way is to just return an observable and use `async` in HTML, but I can see how that might be tricky because you're also potentially setting this property from an https request. The other way to do it would be with `pipe` and `takeUntil(this.ngUnsubscribe)`.

The third way to so save the subscription and unsubscribe, but that one's a hassle :)

File: src/app/home-page/recent-events/recent-events.component.ts
Comment (coverbeck) [inline src/app/home-page/recent-events/recent-events.component.ts:137]: Do you intentionally slice here and intentionally **not** slice on line 117? If no, then create a method to handle both cases to work the same and avoid the duplication.

Another option is do the slicing in the HTML with the Slice pipe. I guess it depends how many events you're expecting; if you expecting hundreds of thousands, then yeah probably best to not keep them all around (and ideally we would handle that in the API so you don't get hundreds of thousands into the browser in the first place). 

I could go either way, and feel free to keep it this way, but I'd lean towards the data being the data, and choosing what portion of it to display being a UI option. Unless, like I said, the data gets to be an unmanageable size.

**Review by svonworl — APPROVED** [review]

File: src/app/home-page/recent-events/recent-events.component.ts
Comment (KimberleyChong) [reply src/app/home-page/recent-events/recent-events.component.ts:137]: Yeah its intentional, I limit to a max of 4 events for the dashboard boxes, but for the user pages we can extend to fit all of the events (returns a max of 10).  It looks like both `selectAll()` and `getEvents()` return a max of 10 events so there's no risk in receiving hundreds.

I'm adding a `displayLimit` member to slice in the HTML and creating a `filterByEventType` method to handle filtering for both cases 👍️

**Review by coverbeck — APPROVED** [review]

**Review by kathy-t — APPROVED** [review]

**Review by denis-yuen — APPROVED** [review]

**Pull request merged**
